### PR TITLE
https://github.com/goadesign/goa/issues/2013 struct:field:type implementation

### PR DESCRIPTION
**Pull request by Danlock**
I also tried to solve a bug in Goa 1 which didn't place the user defined import inside all of the files, just user_types, which led to issues when using this meta tag with Payloads.

Since ImportSpec takes in an additional Name parameter I figured adding qualified imports would be useful as well.

I also noticed the AddImports function wasn't being used anywhere within goa, and made the signature more useful.

### CODE REVIEW

File: codegen/header.go [hunk 1]
```diff
@@ -18,15 +18,18 @@ func Header(title, pack string, imports []*ImportSpec) *SectionTemplate {
 	}
 }
 
-// AddImport adds an import to a section template that was generated with
+// AddImports adds imports to a section template that was generated with
 // Header.
-func AddImport(section *SectionTemplate, imprt *ImportSpec) {
+func AddImports(section *SectionTemplate, imprts ...*ImportSpec) {
```
Comment (nitinmohan87) [inline codegen/header.go]: AddImport is used in quite a few places in goa v2 plugins - https://github.com/goadesign/plugins/blob/master/cors/generate.go#L86

File: codegen/import.go [hunk 2]
```diff
@@ -29,3 +33,82 @@ func (s *ImportSpec) Code() string {
...
+	imports := make([]*ImportSpec, 0)
+	for imp, _ := range uniqueImports {
+		// Copy loop variable into body so next iteration doesnt overwrite its address https://stackoverflow.com/questions/27610039/golang-appending-leaves-only-last-element
+		copy := imp
+		imports = append(imports, &copy)
+	}
+	return imports
+}
+
+func AddServiceMetaTypeImports(header *SectionTemplate, svc *expr.ServiceExpr) {
```
Comment (nitinmohan87) [inline codegen/import.go]: Please add a comment to this exported function.

File: codegen/import.go [hunk 3]
```diff
@@ -29,3 +33,82 @@ func (s *ImportSpec) Code() string {
 	}
 	return fmt.Sprintf(`"%s"`, s.Path)
 }
+
+// GetMetaTypeInfo gets type and import info from an attributes metadata. struct:field:type can have 3 arguments,
```
Comment (nitinmohan87) [inline codegen/import.go]: attributes => attribute's ?

File: codegen/service/service.go [hunk 4]
```diff
@@ -155,6 +154,23 @@ func File(genpkg string, service *expr.ServiceExpr) *codegen.File {
 	return &codegen.File{Path: path, SectionTemplates: sections}
 }
 
+func AddServiceDataMetaTypeImports(header *codegen.SectionTemplate, serviceE *expr.ServiceExpr) {
```
Comment (nitinmohan87) [inline codegen/service/service.go]: Add a comment to this exported function.

File: codegen/generator/example.go [hunk 5]
```diff
@@ -59,6 +59,13 @@ func Example(genpkg string, roots []eval.Root) ([]*codegen.File, error) {
 				files = append(files, fs...)
 			}
 		}
+		for _, f := range files {
+			if len(f.SectionTemplates) > 0 {
+				for _, s := range r.Services {
+					service.AddServiceDataMetaTypeImports(f.SectionTemplates[0], s)
```
Comment (nitinmohan87) [inline codegen/generator/example.go]: Please fix indentation

File: codegen/generator/service.go [hunk 6]
```diff
@@ -26,6 +26,11 @@ func Service(genpkg string, roots []eval.Root) ([]*codegen.File, error) {
 				if f := service.ViewsFile(genpkg, s); f != nil {
 					files = append(files, f)
 				}
+				for _, f := range files {
+					if len(f.SectionTemplates) > 0 {
+						service.AddServiceDataMetaTypeImports(f.SectionTemplates[0], s)
```
Comment (nitinmohan87) [inline codegen/generator/service.go]: Please fix indentation.

File: codegen/generator/transport.go [hunk 7]
```diff
@@ -36,6 +37,14 @@ func Transport(genpkg string, roots []eval.Root) ([]*codegen.File, error) {
 		files = append(files, grpccodegen.ServerTypeFiles(genpkg, r)...)
 		files = append(files, grpccodegen.ClientTypeFiles(genpkg, r)...)
 		files = append(files, grpccodegen.ClientCLIFiles(genpkg, r)...)
+
+		for _, f := range files {
+			if len(f.SectionTemplates) > 0 {
+				for _, s := range r.Services {
+					service.AddServiceDataMetaTypeImports(f.SectionTemplates[0], s)
```
Comment (nitinmohan87) [inline codegen/generator/transport.go]: Please fix indentation.

File: codegen/import.go (hunk 3 above)
Comment (nitinmohan87) [inline codegen/import.go]: Should `GetMetaTypeInfo` be exported? I don't see it being used outside the `codegen` package. If it is not used anywhere else please unexport the function.

File: codegen/generator/example.go (hunk 5 above)
Comment (Danlock) [reply codegen/generator/example.go]: What do you mean? I run gofmt on save, what do you think needs to be fixed?

Comment (nitinmohan87) [reply codegen/generator/example.go]: My bad! Didn't look at the actual file. Github UI confused me a bit.

File: codegen/generator/transport.go (hunk 7 above)
Comment (nitinmohan87) [reply codegen/generator/transport.go]: Indentation looks fine. Sorry!

File: codegen/generator/service.go (hunk 6 above)
Comment (nitinmohan87) [reply codegen/generator/service.go]: Indentation looks fine. Sorry!

File: dsl/meta.go [hunk 8]
```diff
@@ -63,6 +63,22 @@ import (
 //        })
 //    })
 //
+// - "struct:field:type" overrides the Go struct field type specified in the design, with one caveat;
+//		if the type would have been a pointer (such as its not Required) the new type will also be a pointer.
```
Comment (nitinmohan87) [inline dsl/meta.go]: Fix indentation on the entire comment block. Godoc seems to show these comments a bit off.

File: codegen/types_test.go [hunk 9]
```diff
@@ -26,18 +26,23 @@ func TestGoTypeDef(t *testing.T) {
...
-		userType   = &expr.AttributeExpr{Type: ut}
-		resultType = &expr.AttributeExpr{Type: rt}
-		mixedObj   = &expr.AttributeExpr{
+		ut                     = &expr.UserTypeExpr{AttributeExpr: &expr.AttributeExpr{Type: expr.Boolean}, TypeName: "UserType"}
+		rt                     = &expr.ResultTypeExpr{UserTypeExpr: &expr.UserTypeExpr{AttributeExpr: &expr.AttributeExpr{Type: expr.Boolean}, TypeName: "ResultType"}, Identifier: "application/vnd.goa.example", Views: nil}
+		userType               = &expr.AttributeExpr{Type: ut}
+		resultType             = &expr.AttributeExpr{Type: rt}
+		stringMetaType         = expr.MetaExpr{"struct:field:type": []string{"string"}}
+		jsonWithImportMetaType = expr.MetaExpr{"struct:field:type": []string{"json.RawMessage", "encoding/json"}}
+		jsonWithRenameMetaType = expr.MetaExpr{"struct:field:type": []string{"jason.RawMessage", "encoding/json", "jason"}}
```
Comment (nitinmohan87) [inline codegen/types_test.go]: `jason` typo? Or is this intentional?

File: codegen/service/service.go (hunk 4 above)
Comment (nitinmohan87) [reply codegen/service/service.go]: Adds => adds

**Comment by nitinmohan87** [comment]
Just left a few minor fixes. Other than that this looks great!

File: codegen/types_test.go (hunk 9 above)
Comment (nitinmohan87) [reply codegen/types_test.go]: Ah I see. This shows overriding package name?

Comment (Danlock) [reply codegen/types_test.go]: Yes, this is jst testing the 3rd parameter, qualified package

File: codegen/import.go [hunk 10]
```diff
@@ -29,3 +33,83 @@ func (s *ImportSpec) Code() string {
...
+		}
+		if len(args) > 2 {
+			importS.Name = args[2]
+		}
+	}
+	return typeName, importS
+}
+
+// GetMetaTypeImports parses the attribute for all user defined imports
+func GetMetaTypeImports(att *expr.AttributeExpr) []*ImportSpec {
```
Comment (raphael) [inline codegen/import.go]: I believe this function needs to be protected against infinite recursions. Type A may have an attribute of type B which itself has an attribute of type A for example. Take a look at [collectUserTypes](https://github.com/goadesign/goa/blob/v2/http/codegen/service_data.go#L2299) for an example on how to do that. Thank you!

**Comment by raphael** [comment]
@Danlock this looks great - I left a comment in the changes: the function that collects the types defined in metadata should be protected against infinite recursions (see comment for details).

**Comment by raphael** [comment]
Great, thank you!

**Pull request merged**